internal/state/configuration.go

```diff
@@ -87,27 +90,100 @@ func (r *MatchRule) GetMatch() v1beta1.HTTPRouteMatch {
 
 // buildConfiguration builds the Configuration from the graph.
 // FIXME(pleshakov) For now we only handle paths with prefix matches. Handle exact and regex matches
-func buildConfiguration(graph *graph) Configuration {
+func buildConfiguration(
+	ctx context.Context,
+	graph *graph,
+	resolver resolver.ServiceResolver,
+) (Configuration, Warnings) {
 	if graph.GatewayClass == nil || !graph.GatewayClass.Valid {
-		return Configuration{}
+		return Configuration{}, nil
 	}
 
 	if graph.Gateway == nil {
-		return Configuration{}
+		return Configuration{}, nil
 	}
 
-	upstreams := buildUpstreams(graph.Gateway.Listeners)
+	upstreamsMap := buildUpstreamsMap(ctx, graph.Gateway.Listeners, resolver)
 	httpServers, sslServers := buildServers(graph.Gateway.Listeners)
 	backendGroups := buildBackendGroups(graph.Gateway.Listeners)
 
+	warnings := buildWarnings(graph, upstreamsMap)
+
 	config := Configuration{
 		HTTPServers:   httpServers,
 		SSLServers:    sslServers,
-		Upstreams:     upstreams,
+		Upstreams:     upstreamsMapToSlice(upstreamsMap),
 		BackendGroups: backendGroups,
 	}
 
-	return config
+	return config, warnings
+}
+
+func upstreamsMapToSlice(upstreamsMap map[string]Upstream) []Upstream {
+	if len(upstreamsMap) == 0 {
+		return nil
+	}
+
+	upstreams := make([]Upstream, 0, len(upstreamsMap))
+	for _, upstream := range upstreamsMap {
+		upstreams = append(upstreams, upstream)
+	}
+
+	return upstreams
+}
+
+func buildWarnings(graph *graph, upstreams map[string]Upstream) Warnings {
+	warnings := newWarnings()
+
+	for _, l := range graph.Gateway.Listeners {
+		for _, r := range l.Routes {
+			if !l.Valid {
+				warnings.AddWarningf(
+					r.Source,
+					"cannot configure routes for listener %s; listener is invalid",
+					l.Source.Name,
+				)
+
+				continue
+			}
+
+			for _, group := range r.BackendGroups {
+
+				for _, errMsg := range group.Errors {
+					warnings.AddWarningf(r.Source, "invalid backend ref: %s", errMsg)
+				}
+
+				for _, backend := range group.Backends {
+					if backend.Name != "" {
+						upstream, ok := upstreams[backend.Name]
+
+						// this should never happen; but we check it just in case
+						if !ok {
+							warnings.AddWarningf(
+								r.Source,
+								"invalid backend ref: upstream %s not found in map",
```

Comment:
I'm fine with a panic here. @f5yacobucci thoughts?

Reply:
For now, leave this as a warning, but add Internal Server indication. I would like to write up best practices for use of panic - I think this is too deep into the runtime and too late to panic. An otherwise running dataplane can be impacted negatively.